api/src/main/scala/dx/api/DxApi.scala

```diff
@@ -951,23 +951,35 @@ case class DxApi(version: String = "1.0.0", dxEnv: DXEnvironment = DxApi.default
       }
 
     val allResults = workspaceResults ++ remaining.groupBy(_.project).flatMap {
-      case (None, files) if !inExecutionEnvironment && currentProjectId.isDefined =>
+      case (None, files) if currentProjectId.isDefined =>
```

Comment:
In dxCompiler we always use `true` for `searchWorkspaceFirst`, but that will make unnecessary api search if the files are in different projects (aka their projectID are set but here we first look them up in the workspace)...
https://github.com/search?q=org%3Adnanexus+searchWorkspaceFirst&type=code

Also here could that ever be possible that `currentWorkspaceId.isDefined` but `currentProjectId` is not? I saw `currentWorkspaceID` is firstly from the dxEnv then the currentProjectID if not in a job exec env

Reply:
> In dxCompiler we always use `true` for `searchWorkspaceFirst`, but that will make unnecessary api search if the files are in different projects (aka their projectID are set but here we first look them up in the workspace)... https://github.com/search?q=org%3Adnanexus+searchWorkspaceFirst&type=code
> 

Good point. But what if there was an operation upstream which already cloned a given `file-xxx` from the `project-aaa` into the execution `container-kkk`? And then there's a downstream operation that clones objects based on their description (whether they are in the exec `container-kkk`). We don't want to miss that already cloned `file-xxx` and potentially clone it multiple times over the execution. Another point - this is a method for a bulk description (up to 1k IDs) whether we describe the same `file-xxx` multiple times - is a matter of just a few additional calls. E.g. you have 10k files in an operation. Even if all of them were already cloned - that's a matter of 10 API calls. I guess that's the scale when calls for cloning might be a tighter bottleneck. 

> Also here could that ever be possible that `currentWorkspaceId.isDefined` but `currentProjectId` is not? I saw `currentWorkspaceID` is firstly from the dxEnv then the currentProjectID if not in a job exec env

Maybe I just don't know of such a corner case, but I can't imagine an execution container being outside of a project. [Here](https://github.com/dnanexus/dxScala/blob/7ac0ca7222b89d040f8f67d792eedde79970fec8/api/src/main/scala/dx/api/DxApi.scala#L117-L118)